#### Open and read the file `Hawking-Questioning-the-Universe.txt`.

In [99]:
with open("Hawking-Questioning-the-Universe.txt", "r") as file:
    text = file.read()

#### Print the first 50 characters.

In [3]:
print(text[0:50])

There is nothing bigger or older than the universe


#### Print from the 52nd to the 163rd character.

In [191]:
print(text[51:163])

 The questions I would like to talk about are: one, where did we come from? How did the universe come into bein


---

#### Open the file `Hawking-Questioning-the-Universe.txt` and create a list named `text_as_list` where every word is an element of the list. 

Hint: there are two ways to do it. Either with `.readlines()` or with `.split()`. I encourage you to google how both work, but I suggest to adopt the latter.

Hint: if you want to split the text so every word is an element of a list, what should be the parameter of the `split()` method? Read more [here](https://www.w3schools.com/python/ref_string_split.asp).

In [111]:
with open("Hawking-Questioning-the-Universe.txt", "r") as file:
    text = file.read()
text_as_list = text.split()

#### This code will remove the punctuation from your list and applies the lower() method that returns a string where all characters are lower case. Simple run the cell.

In [112]:
import string
text_as_list = [i.translate(str.maketrans('', '', string.punctuation)).lower() for i in text_as_list]

#### Create a new list `words_begin_with_a` and append all words of the `text_as_list` that begin with a small _a_.

Hint: you have to define first an empty list.

Hint: in order to find if a string starts with one or more letters, you should use the method `.startswith()`. Read more [here](https://www.w3schools.com/python/ref_string_startswith.asp).

In [67]:
words_begin_with_a = []
for item in text_as_list:
    if item.startswith("a"):
        words_begin_with_a.append(item)
    else:
        continue
words_begin_with_a

['about',
 'are:',
 'alone',
 'alien',
 'and',
 'away',
 'all',
 'about',
 'ago.',
 'anything',
 'and',
 'all',
 'at',
 'and',
 'all',
 'about',
 'and',
 'and',
 'and',
 'and',
 'allow',
 'another',
 'and',
 'and',
 'also',
 'a',
 'are',
 'agreement',
 'an',
 'and',
 'are',
 'alone,',
 'arose',
 'appear',
 'a',
 'appeared.',
 'appearing.',
 'algae',
 'ago.',
 'ago',
 'and',
 'about',
 'appeared',
 'a',
 'a',
 'appearing',
 'available.',
 'aliens.',
 'am',
 'appear',
 'and',
 'a',
 'and',
 'aliens',
 'a',
 'an',
 'any',
 'alien',
 'are',
 'alien',
 'at',
 'a',
 'a',
 'an',
 'against',
 'abduction',
 'aliens',
 'a',
 'are',
 'and',
 'are',
 'an',
 'and',
 'are',
 'along',
 'ability',
 'and',
 'aggressive',
 'advantage',
 'avoid',
 'alone',
 'answers',
 'am',
 'and',
 'answers',
 'a',
 'a',
 'and',
 'are']

Answer: 310

#### Create a counter and use a while loop to counter how many words starts with a _u_.

In [86]:
i = 0
counter = 0

while i<len(text_as_list):
    if text_as_list[i].startswith('u'):
        counter += 1
    i += 1
print(counter)

30


Answer: 30

#### Create a counter and use a while loop to counter how many words do NOT start with a _u_.

In [87]:
i = 0
counter2 = 0

while i<len(text_as_list):
    if not text_as_list[i].startswith('u'):
        counter2 += 1
    i += 1
print(counter2)

877


Answer: 877

#### Write an if statement that checkes if the word "universe" appears in the text. Feel free to google how we test if a word is _in_ a Python list.

In [113]:
i = text_as_list.count("universe")
if i > 0:
    print("The word does appear")
else:
    print("The word does not appear")

The word does appear


In [119]:
# checklist1 = ['universe']
# flag = 0
# for i in text_as_list:
#     for j in checklist1:
#         if i==j:
#             flag += 1
#             break
# if flag == 1:
#     print("The word does appear")
# else:
#     print("The word does not appear")

#### Write an if statement that checkes if the word "alliens" appears in the text. If not print "the word does not appear"

In [115]:
i = text_as_list.count("alliens")
if i > 0:
    print("The word does appear")
else:
    print("The word does not appear")

The word does not appear


In [118]:
# checklist2 = ["alliens"]
# flag = 0
# for i in text:
#     for j in checklist2:
#         if i == j:
#             flag += 1
#             break
# if flag == 1:
#     print("The word does appear")
# else:
#     print("The word does not appear")

#### Write an if statement that checkes if the word "alliens" appears in the text. If not, check if the word "International Space Station" appears in the text. If neither is in the text, print "I didn't find the words".

In [116]:
i = text_as_list.count("alliens")
j = text_as_list.count("International Space Station")
if i > 0:
    print("The word does appear")
elif j > 0:
    print("The word does appear")
else:
    print("I didn't find the words")

I didn't find the words


In [117]:
# checklist3 = ["alliens", "International Space Station"]
# flag1 = 0
# flag2 = 0
# for i in text_as_list:
#     for j in checklist3:
#         if i == j[0]:
#             flag1 += 1
#         elif i == j[1]:
#             flag2 += 1
#             break
# if flag1 == 1:
#     print("The word does appear")
# elif flag2 == 1:
#     print("The word does appear")
# else:
#     print("I didn't find the words")

#### Take the text_as_list and create a dictionary named `mydictionary`. The words of the text should be the keys of the dictionary and the index of the word the values of the keys.

Hint: there is a method in Python that adds a counter to an iterable.

In [192]:
mydictionary = {}
def my_enumerate(sequence, start=1):
    n = start
    for elem in sequence:
        yield n, elem
        n += 1
my_enumerate(text_as_list)
mydictionary = dict(my_enumerate(text_as_list))

#### Create the same dictonary but now index of  words in the text should be the dictionary's keys and the words should be the key's values.

In [145]:
newdictionary = {}
for key, value in mydictionary.items():
    if i:
        newdictionary[value]=[key]
print(len(newdictionary))

358


#### Why the lenghts of the two dictionaries are not the same?

The new dictionary counts the number of instances that each unique value occurs while the old dictionary counts the number of unique instances that each value occurs

# Create a dictionary `dict_counter` that takes as keys the words of text_as_list. The values of each key should be the number that each words is attested in the list.

Hint: Check the [`count()` method](https://www.w3schools.com/python/ref_list_count.asp).

In [188]:
dict_counter = {}
for i in text_as_list:
    dict_counter[i] = text_as_list.count(i)
dict_counter

{'there': 10,
 'is': 13,
 'nothing': 2,
 'bigger': 1,
 'or': 5,
 'older': 1,
 'than': 2,
 'the': 77,
 'universe': 19,
 'questions': 4,
 'i': 7,
 'would': 3,
 'like': 3,
 'to': 22,
 'talk': 1,
 'about': 4,
 'are': 9,
 'one': 3,
 'where': 1,
 'did': 4,
 'we': 26,
 'come': 2,
 'from': 4,
 'how': 2,
 'into': 4,
 'being': 3,
 'alone': 3,
 'in': 14,
 'alien': 3,
 'life': 9,
 'out': 3,
 'what': 2,
 'future': 3,
 'of': 39,
 'human': 2,
 'race': 2,
 'up': 1,
 'until': 2,
 '1920s': 1,
 'everyone': 2,
 'thought': 1,
 'was': 10,
 'essentially': 1,
 'static': 1,
 'and': 19,
 'unchanging': 1,
 'time': 6,
 'then': 1,
 'it': 9,
 'discovered': 1,
 'that': 12,
 'expanding': 1,
 'distant': 1,
 'galaxies': 1,
 'were': 3,
 'moving': 1,
 'away': 1,
 'us': 1,
 'this': 7,
 'meant': 1,
 'they': 2,
 'must': 3,
 'have': 14,
 'been': 6,
 'closer': 1,
 'together': 1,
 'past': 2,
 'if': 6,
 'extrapolate': 1,
 'back': 1,
 'find': 2,
 'all': 4,
 'on': 9,
 'top': 1,
 'each': 1,
 'other': 4,
 '15': 1,
 'billion': 6,
 '

In [ ]:
dict_counter = {}
for key, value in mydictionary.items():
    if i:
        dict_counter [value]=[key]
    for i in text_as_list:
        text_as_list.count(i)
        n += 1
my_enumerate(text_as_list)
dict_counter = dict((text_as_list))
print(len(mydictionary))

#### Without printing the whole dictionary, find how many times the word "question" appears in the list by using your new dictiornary

In [149]:
i = text_as_list.count("question")
print(i)

2


#### Create a pandas dataframe from the items of the dictionary

In [154]:
import pandas as pd
df = pd.DataFrame.from_dict(mydictionary)
df

ValueError: If using all scalar values, you must pass an index

#### Get the first row of your dataframe

In [ ]:
df.head(1)

#### Get the first column of your dataframe

In [153]:
df.iloc[:,0]

NameError: name 'df' is not defined

#### By using the `pandas.DataFrame.loc`, find how many times the words galaxies and galaxies appear in the dataframe

#### Write a recursion in a function that prints 5! (i.e., the factorial of 5).

In [179]:
def fact(n):
    if n == 1:
        return n
    else:
        return n*fact(n-1)
n = 5

In [180]:
print(fact(n))

120


#### Write a recursion in a function that prints the $n^{th}$ [Fibonacci number](https://en.wikipedia.org/wiki/Fibonacci_number):

1, 1, 2, 3, 5, 8, 13, 21, 34

In [189]:
def fib(n):
    if n <=1:
        return n
    else:
        return fib(n-1) + fib(n-2)

#### Which is the $20^{th}$ Fibonacci number?

In [190]:
fib(20)

6765

_We will try together tail recursion._